File: buttons.js
// buttons.js - Button assets management and fallback UI system

/**
 * Manages button assets and provides fallback rendering for UI buttons.
 * This class handles loading button images and provides fallback styling
 * when assets are not available.
 */
export class ButtonAssets {
  constructor(assetsPath = 'assets/buttons/') {
    this.assetsPath = assetsPath;
    this.buttonAssets = new Map();
    this.fallbackStyles = new Map();
    this.loadedImages = new Map();
    
    // Initialize button asset mappings
    this._initButtonAssets();
    
    // Initialize fallback styles
    this._initFallbackStyles();
  }
  
  _initButtonAssets() {
    // Initialize the mapping of button names to asset file paths
    this.buttonAssets.set('shop', `${this.assetsPath}shop.png`);
    this.buttonAssets.set('start', `${this.assetsPath}start.png`);
    this.buttonAssets.set('load', `${this.assetsPath}load.png`);
    this.buttonAssets.set('save', `${this.assetsPath}save.png`);
    this.buttonAssets.set('resume', `${this.assetsPath}resume.png`);
    this.buttonAssets.set('quit', `${this.assetsPath}quit.png`);
    this.buttonAssets.set('rivals', `${this.assetsPath}rivals.png`);
    this.buttonAssets.set('settings', `${this.assetsPath}settings.png`);
    this.buttonAssets.set('back', `${this.assetsPath}back.png`);
    this.buttonAssets.set('close', `${this.assetsPath}close.png`);
  }
  
  _initFallbackStyles() {
    // Initialize fallback styling for buttons when assets are not available
    this.fallbackStyles.set('shop', {
      bgColor: '#FFD700',        // Gold
      textColor: '#000000',
      borderColor: '#FFA500',
      hoverBgColor: '#FFC107',
      text: 'SHOP',
      fontSize: 16,
      borderWidth: 2,
      borderRadius: 8
    });
    
    this.fallbackStyles.set('start', {
      bgColor: '#28A745',        // Green
      textColor: '#FFFFFF',
      borderColor: '#1E7E34',
      hoverBgColor: '#218838',
      text: 'START',
      fontSize: 16,
      borderWidth: 2,
      borderRadius: 8
    });
    
    this.fallbackStyles.set('load', {
      bgColor: '#17A2B8',        // Cyan
      textColor: '#FFFFFF',
      borderColor: '#138496',
      hoverBgColor: '#148A9C',
      text: 'LOAD',
      fontSize: 16,
      borderWidth: 2,
      borderRadius: 8
    });
    
    this.fallbackStyles.set('save', {
      bgColor: '#6F42C1',        // Purple
      textColor: '#FFFFFF',
      borderColor: '#5A2D91',
      hoverBgColor: '#5D2E8B',
      text: 'SAVE',
      fontSize: 16,
      borderWidth: 2,
      borderRadius: 8
    });
    
    this.fallbackStyles.set('resume', {
      bgColor: '#2563EB',        // Blue
      textColor: '#FFFFFF',
      borderColor: '#1E40AF',
      hoverBgColor: '#1D4ED8',
      text: 'RESUME',
      fontSize: 16,
      borderWidth: 2,
      borderRadius: 8
    });
    
    this.fallbackStyles.set('quit', {
      bgColor: '#DC3545',        // Red
      textColor: '#FFFFFF',
      borderColor: '#C82333',
      hoverBgColor: '#BD2130',
      text: 'QUIT',
      fontSize: 16,
      borderWidth: 2,
      borderRadius: 8
    });
    
    this.fallbackStyles.set('rivals', {
      bgColor: '#FD7E14',        // Orange
      textColor: '#FFFFFF',
      borderColor: '#DC6A00',
      hoverBgColor: '#E8681F',
      text: 'RIVALS',
      fontSize: 16,
      borderWidth: 2,
      borderRadius: 8
    });
    
    this.fallbackStyles.set('settings', {
      bgColor: '#6C757D',        // Gray
      textColor: '#FFFFFF',
      borderColor: '#5A6268',
      hoverBgColor: '#5A6268',
      text: 'SETTINGS',
      fontSize: 16,
      borderWidth: 2,
      borderRadius: 8
    });
    
    this.fallbackStyles.set('back', {
      bgColor: '#6C757D',        // Gray
      textColor: '#FFFFFF',
      borderColor: '#5A6268',
      hoverBgColor: '#5A6268',
      text: 'BACK',
      fontSize: 16,
      borderWidth: 2,
      borderRadius: 8
    });
    
    this.fallbackStyles.set('close', {
      bgColor: '#DC3545',        // Red
      textColor: '#FFFFFF',
      borderColor: '#C82333',
      hoverBgColor: '#BD2130',
      text: 'X',
      fontSize: 18,
      borderWidth: 2,
      borderRadius: 4
    });
  }
  
  /**
   * Get the file path for a button asset
   * @param {string} buttonName - Name of the button (e.g., 'shop', 'start', 'load')
   * @returns {string|null} File path to the button asset, or null if not found
   */
  getButtonAssetPath(buttonName) {
    return this.buttonAssets.get(buttonName.toLowerCase()) || null;
  }
  
  /**
   * Load a button image asset
   * @param {string} buttonName - Name of the button
   * @returns {Promise<HTMLImageElement>} Promise that resolves to the loaded image
   */
  async loadButtonAsset(buttonName) {
    const key = buttonName.toLowerCase();
    
    // Return cached image if already loaded
    if (this.loadedImages.has(key)) {
      return this.loadedImages.get(key);
    }
    
    const assetPath = this.getButtonAssetPath(buttonName);
    if (!assetPath) {
      throw new Error(`No asset path found for button: ${buttonName}`);
    }
    
    return new Promise((resolve, reject) => {
      const img = new Image();
      
      img.onload = () => {
        this.loadedImages.set(key, img);
        resolve(img);
      };
      
      img.onerror = () => {
        reject(new Error(`Failed to load button asset: ${assetPath}`));
      };
      
      img.src = assetPath;
    });
  }
  
  /**
   * Check if a button image is already loaded
   * @param {string} buttonName - Name of the button
   * @returns {boolean} True if the image is loaded
   */
  isLoaded(buttonName) {
    return this.loadedImages.has(buttonName.toLowerCase());
  }
  
  /**
   * Get a loaded button image
   * @param {string} buttonName - Name of the button
   * @returns {HTMLImageElement|null} The loaded image or null if not loaded
   */
  getLoadedImage(buttonName) {
    return this.loadedImages.get(buttonName.toLowerCase()) || null;
  }
  
  /**
   * Get fallback styling for a button
   * @param {string} buttonName - Name of the button
   * @returns {Object} Fallback style properties
   */
  getFallbackStyle(buttonName) {
    return this.fallbackStyles.get(buttonName.toLowerCase()) || {
      bgColor: '#2563EB',
      textColor: '#FFFFFF',
      borderColor: '#1E40AF',
      hoverBgColor: '#1D4ED8',
      text: buttonName.toUpperCase(),
      fontSize: 16,
      borderWidth: 2,
      borderRadius: 8
    };
  }
  
  /**
   * Get complete button configuration including asset path and fallback
   * @param {string} buttonName - Name of the button
   * @returns {Object} Button configuration object
   */
  getButtonConfig(buttonName) {
    return {
      name: buttonName,
      assetPath: this.getButtonAssetPath(buttonName),
      isLoaded: this.isLoaded(buttonName),
      image: this.getLoadedImage(buttonName),
      fallbackStyle: this.getFallbackStyle(buttonName)
    };
  }
    /**
   * Draw a button using either the asset image or fallback styling
   * @param {CanvasRenderingContext2D} ctx - Canvas context
   * @param {string} buttonName - Name of the button
   * @param {number} x - X position
   * @param {number} y - Y position
   * @param {number} width - Button width
   * @param {number} height - Button height
   * @param {boolean} isHovered - Whether the button is being hovered
   * @param {number} scale - Scale factor for text and border elements (default: 1)
   * @returns {Object} Button bounds for click detection
   */  drawButton(ctx, buttonName, x, y, width, height, isHovered = false, scale = 1) {
    const config = this.getButtonConfig(buttonName);
    
    // Try to draw with image asset first
    if (config.image && config.image.complete && config.image.naturalWidth > 0) {
      // Draw the image asset
      ctx.drawImage(config.image, x, y, width, height);
    } else {
      // Fall back to styled button
      const style = config.fallbackStyle;
      const bgColor = isHovered ? style.hoverBgColor : style.bgColor;
      
      // Scale border radius and line width with better bounds
      const scaledBorderRadius = Math.max(2, Math.min(15, style.borderRadius * scale));
      const scaledBorderWidth = Math.max(1, Math.min(4, style.borderWidth * scale));
      
      // Draw button background
      ctx.fillStyle = bgColor;
      this._drawRoundedRect(ctx, x, y, width, height, scaledBorderRadius);
      ctx.fill();
      
      // Draw button border
      ctx.strokeStyle = style.borderColor;
      ctx.lineWidth = scaledBorderWidth;
      this._drawRoundedRect(ctx, x, y, width, height, scaledBorderRadius);
      ctx.stroke();
      
      // Draw button text with scaled font size and better bounds
      ctx.fillStyle = style.textColor;
      const baseFontSize = Math.min(style.fontSize, height * 0.5); // Don't exceed half button height
      const scaledFontSize = Math.max(8, Math.min(24, baseFontSize * scale));
      ctx.font = `bold ${scaledFontSize}px sans-serif`;
      ctx.textAlign = 'center';
      ctx.textBaseline = 'middle';
      ctx.fillText(style.text, x + width / 2, y + height / 2);
    }
    
    // Return button bounds for click detection
    return { x, y, w: width, h: height, action: buttonName };
  }
  
  /**
   * Helper method to draw rounded rectangles
   * @private
   */
  _drawRoundedRect(ctx, x, y, w, h, r) {
    ctx.beginPath();
    ctx.moveTo(x + r, y);
    ctx.lineTo(x + w - r, y);
    ctx.quadraticCurveTo(x + w, y, x + w, y + r);
    ctx.lineTo(x + w, y + h - r);
    ctx.quadraticCurveTo(x + w, y + h, x + w - r, y + h);
    ctx.lineTo(x + r, y + h);
    ctx.quadraticCurveTo(x, y + h, x, y + h - r);
    ctx.lineTo(x, y + r);
    ctx.quadraticCurveTo(x, y, x + r, y);
    ctx.closePath();
  }
  
  /**
   * Add a custom button configuration
   * @param {string} buttonName - Name of the button
   * @param {string} assetPath - Path to the button asset (optional)
   * @param {Object} fallbackStyle - Fallback style object (optional)
   */
  addCustomButton(buttonName, assetPath = null, fallbackStyle = null) {
    const key = buttonName.toLowerCase();
    
    if (assetPath) {
      this.buttonAssets.set(key, assetPath);
    }
    
    if (fallbackStyle) {
      this.fallbackStyles.set(key, fallbackStyle);
    }
  }
  
  /**
   * Preload all button assets
   * @returns {Promise<void>} Promise that resolves when all assets are loaded
   */
  async preloadAll() {
    const loadPromises = [];
    
    for (const buttonName of this.buttonAssets.keys()) {
      loadPromises.push(
        this.loadButtonAsset(buttonName).catch(err => {
          console.warn(`Failed to load button asset for ${buttonName}:`, err.message);
          return null; // Continue with fallback
        })
      );
    }
    
    await Promise.all(loadPromises);
  }
  
  /**
   * Get a list of all available button names
   * @returns {string[]} Array of button names
   */
  listAvailableButtons() {
    return Array.from(this.buttonAssets.keys());
  }
}

// Global instance for easy access
export const buttonAssets = new ButtonAssets();

// Preload button assets when module is imported
buttonAssets.preloadAll().catch(err => {
  console.warn('Some button assets failed to preload:', err);
});

File: buttons.py
# buttons.py - Button assets management and fallback UI system

import os
from typing import Dict, Optional, Tuple

class ButtonAssets:
    """
    Manages button assets and provides fallback rendering for UI buttons.
    This class handles loading button images and provides fallback styling
    when assets are not available.
    """
    
    def __init__(self, assets_path: str = "assets/buttons/"):
        self.assets_path = assets_path
        self.button_assets: Dict[str, str] = {}
        self.fallback_styles: Dict[str, Dict] = {}
        
        # Initialize button asset mappings
        self._init_button_assets()
        
        # Initialize fallback styles
        self._init_fallback_styles()
    
    def _init_button_assets(self):
        """Initialize the mapping of button names to asset file paths."""
        self.button_assets = {
            'shop': os.path.join(self.assets_path, 'shop.png'),
            'start': os.path.join(self.assets_path, 'start.png'),
            'load': os.path.join(self.assets_path, 'load.png'),
            'save': os.path.join(self.assets_path, 'save.png'),
            'resume': os.path.join(self.assets_path, 'resume.png'),
            'quit': os.path.join(self.assets_path, 'quit.png'),
            'rivals': os.path.join(self.assets_path, 'rivals.png'),
            'settings': os.path.join(self.assets_path, 'settings.png'),
            'back': os.path.join(self.assets_path, 'back.png'),
            'close': os.path.join(self.assets_path, 'close.png'),
        }
    
    def _init_fallback_styles(self):
        """Initialize fallback styling for buttons when assets are not available."""
        self.fallback_styles = {
            'shop': {
                'bg_color': '#FFD700',  # Gold
                'text_color': '#000000',
                'border_color': '#FFA500',
                'hover_bg_color': '#FFC107',
                'text': 'SHOP',
                'font_size': 16,
                'border_width': 2,
                'border_radius': 8
            },
            'start': {
                'bg_color': '#28A745',  # Green
                'text_color': '#FFFFFF',
                'border_color': '#1E7E34',
                'hover_bg_color': '#218838',
                'text': 'START',
                'font_size': 16,
                'border_width': 2,
                'border_radius': 8
            },
            'load': {
                'bg_color': '#17A2B8',  # Cyan
                'text_color': '#FFFFFF',
                'border_color': '#138496',
                'hover_bg_color': '#148A9C',
                'text': 'LOAD',
                'font_size': 16,
                'border_width': 2,
                'border_radius': 8
            },
            'save': {
                'bg_color': '#6F42C1',  # Purple
                'text_color': '#FFFFFF',
                'border_color': '#5A2D91',
                'hover_bg_color': '#5D2E8B',
                'text': 'SAVE',
                'font_size': 16,
                'border_width': 2,
                'border_radius': 8
            },
            'resume': {
                'bg_color': '#2563EB',  # Blue
                'text_color': '#FFFFFF',
                'border_color': '#1E40AF',
                'hover_bg_color': '#1D4ED8',
                'text': 'RESUME',
                'font_size': 16,
                'border_width': 2,
                'border_radius': 8
            },
            'quit': {
                'bg_color': '#DC3545',  # Red
                'text_color': '#FFFFFF',
                'border_color': '#C82333',
                'hover_bg_color': '#BD2130',
                'text': 'QUIT',
                'font_size': 16,
                'border_width': 2,
                'border_radius': 8
            },
            'rivals': {
                'bg_color': '#FD7E14',  # Orange
                'text_color': '#FFFFFF',
                'border_color': '#DC6A00',
                'hover_bg_color': '#E8681F',
                'text': 'RIVALS',
                'font_size': 16,
                'border_width': 2,
                'border_radius': 8
            },
            'settings': {
                'bg_color': '#6C757D',  # Gray
                'text_color': '#FFFFFF',
                'border_color': '#5A6268',
                'hover_bg_color': '#5A6268',
                'text': 'SETTINGS',
                'font_size': 16,
                'border_width': 2,
                'border_radius': 8
            },
            'back': {
                'bg_color': '#6C757D',  # Gray
                'text_color': '#FFFFFF',
                'border_color': '#5A6268',
                'hover_bg_color': '#5A6268',
                'text': 'BACK',
                'font_size': 16,
                'border_width': 2,
                'border_radius': 8
            },
            'close': {
                'bg_color': '#DC3545',  # Red
                'text_color': '#FFFFFF',
                'border_color': '#C82333',
                'hover_bg_color': '#BD2130',
                'text': 'X',
                'font_size': 18,
                'border_width': 2,
                'border_radius': 4
            }
        }
    
    def get_button_asset_path(self, button_name: str) -> Optional[str]:
        """
        Get the file path for a button asset.
        
        Args:
            button_name: Name of the button (e.g., 'shop', 'start', 'load')
            
        Returns:
            File path to the button asset, or None if not found
        """
        return self.button_assets.get(button_name.lower())
    
    def has_asset(self, button_name: str) -> bool:
        """
        Check if a button asset file exists.
        
        Args:
            button_name: Name of the button
            
        Returns:
            True if the asset file exists, False otherwise
        """
        asset_path = self.get_button_asset_path(button_name)
        return asset_path is not None and os.path.exists(asset_path)
    
    def get_fallback_style(self, button_name: str) -> Dict:
        """
        Get fallback styling for a button.
        
        Args:
            button_name: Name of the button
            
        Returns:
            Dictionary containing fallback style properties
        """
        return self.fallback_styles.get(button_name.lower(), {
            'bg_color': '#2563EB',
            'text_color': '#FFFFFF',
            'border_color': '#1E40AF',
            'hover_bg_color': '#1D4ED8',
            'text': button_name.upper(),
            'font_size': 16,
            'border_width': 2,
            'border_radius': 8
        })
    
    def get_button_config(self, button_name: str) -> Dict:
        """
        Get complete button configuration including asset path and fallback.
        
        Args:
            button_name: Name of the button
            
        Returns:
            Dictionary containing asset path (if available) and fallback style
        """
        config = {
            'name': button_name,
            'asset_path': self.get_button_asset_path(button_name),
            'has_asset': self.has_asset(button_name),
            'fallback_style': self.get_fallback_style(button_name)
        }
        return config
    
    def add_custom_button(self, button_name: str, asset_path: str = None, fallback_style: Dict = None):
        """
        Add a custom button configuration.
        
        Args:
            button_name: Name of the button
            asset_path: Path to the button asset (optional)
            fallback_style: Fallback style dictionary (optional)
        """
        if asset_path:
            self.button_assets[button_name.lower()] = asset_path
        
        if fallback_style:
            self.fallback_styles[button_name.lower()] = fallback_style
    
    def list_available_buttons(self) -> list:
        """
        Get a list of all available button names.
        
        Returns:
            List of button names
        """
        return list(self.button_assets.keys())
    
    def get_asset_dimensions(self, button_name: str) -> Optional[Tuple[int, int]]:
        """
        Get the dimensions of a button asset (if available).
        This is a placeholder - actual implementation would require image processing.
        
        Args:
            button_name: Name of the button
            
        Returns:
            Tuple of (width, height) or None if asset not available
        """
        # This would require PIL or similar library to actually read image dimensions
        # For now, return standard button dimensions
        if self.has_asset(button_name):
            return (160, 40)  # Standard button size
        return None


# Global instance for easy access
button_assets = ButtonAssets()

# Export main functionality
__all__ = ['ButtonAssets', 'button_assets']


# Example usage:
if __name__ == "__main__":
    # Example of how to use the ButtonAssets class
    assets = ButtonAssets()
    
    # Check if shop button asset exists
    shop_config = assets.get_button_config('shop')
    print(f"Shop button config: {shop_config}")
    
    # List all available buttons
    print(f"Available buttons: {assets.list_available_buttons()}")
    
    # Add a custom button
    assets.add_custom_button('inventory', 'assets/buttons/inventory.png', {
        'bg_color': '#8B4513',
        'text_color': '#FFFFFF',
        'text': 'INVENTORY'
    })
